Outbox: record logs

mattwiebe opened the pull request:
We don't explicitly have to merge this, since we're already two levels deep on a feature branch, but this is the idea.

We can build on this with CLI and in the Outbox post listing, including retry functionality, and "copy to clipboard" so users can submit that with bug reports.

pfefferle committed:
init

pfefferle committed:
outbox should not be public!

pfefferle committed:
add basic `to_activity` function

pfefferle committed:
Merge branch 'master' into add/outbox-collection

pfefferle committed:
do not allow to instance post transformer

pfefferle committed:
Merge branch 'master' into add/outbox-collection

pfefferle committed:
fix tests

pfefferle committed:
Merge branch 'master' into add/outbox-collection

pfefferle committed:
Merge branch 'master' into add/outbox-collection

pfefferle committed:
Merge branch 'master' into add/outbox-collection

pfefferle committed:
Merge branch 'trunk' into add/outbox-collection

pfefferle committed:
fix phpcs

pfefferle committed:
revert change

pfefferle committed:
remove unused `use` declarations

pfefferle committed:
the handler should not handle outgoing stuff

pfefferle committed:
Merge branch 'trunk' into add/outbox-collection

pfefferle committed:
Update class-activitypub.php

pfefferle committed:
fix PHPCS

pfefferle committed:
more PHPCS fixes

pfefferle committed:
fix namespace issue

pfefferle committed:
remove unneeded function

pfefferle committed:
fix sticky post endpoint

/cc @mattwiebe

pfefferle committed:
no need to check for User-ID

pfefferle committed:
Merge branch 'trunk' into add/outbox-collection

pfefferle committed:
Merge branch 'trunk' into add/outbox-collection

pfefferle committed:
Merge branch 'trunk' into add/outbox-collection

pfefferle committed:
Merge branch 'trunk' into add/outbox-collection

pfefferle committed:
Merge branch 'trunk' into add/outbox-collection

pfefferle committed:
Merge branch 'trunk' into add/outbox-collection

pfefferle committed:
support JSON and Arrays beside WP_Comments and WP_Posts

pfefferle committed:
fix auto-complete issue

pfefferle committed:
simplify code.

pfefferle committed:
convert JSON to Activity_Object to not have everything duplicated

pfefferle committed:
add support for maps and mentions

pfefferle committed:
Merge branch 'trunk' into add/outbox-collection

pfefferle committed:
it takes a string (JSON) or an array

pfefferle committed:
Merge branch 'trunk' into add/outbox-collection

pfefferle committed:
simplified based on the feedback of @obenland

pfefferle committed:
Merge branch 'add/outbox-collection' of https://github.com/Automattic/wordpress-activitypub into add/outbox-collection

pfefferle committed:
rearrange code a bit

pfefferle committed:
add missing uses

pfefferle committed:
fix typo

pfefferle committed:
add missing uses

pfefferle committed:
Merge branch 'trunk' into add/outbox-collection

pfefferle committed:
Merge branch 'trunk' into add/outbox-collection

pfefferle committed:
Merge branch 'trunk' into add/outbox-collection

pfefferle committed:
Merge branch 'trunk' into add/outbox-collection

pfefferle committed:
Merge branch 'trunk' into add/outbox-collection

pfefferle committed:
Merge branch 'trunk' into add/outbox-collection

pfefferle committed:
Added Changelog-Entry

pfefferle committed:
Merge branch 'trunk' into add/outbox-collection

pfefferle committed:
apply changes from @obenland

pfefferle committed:
delay scheduling a bit

pfefferle committed:
change scheduling and add possibility to add items to outbox

pfefferle committed:
remove deprecated functions

pfefferle committed:
add filter

pfefferle committed:
Merge branch 'trunk' into add/outbox-collection

pfefferle committed:
fix indents

pfefferle committed:
removed deprecated functions

pfefferle committed:
add visibility and normalize autor id

pfefferle committed:
add taxonomies to store actor and actitiy-type informations

pfefferle committed:
Merge branch 'trunk' into add/outbox-collection

pfefferle committed:
Merge branch 'trunk' into add/outbox-collection

pfefferle committed:
Merge branch 'trunk' into add/outbox-collection

obenland committed:
Merge branch 'trunk' into add/outbox-collection

obenland committed:
Fix phpcs

obenland committed:
Fix shortcode tests

obenland committed:
Update json tests.

pfefferle committed:
Merge branch 'trunk' into add/outbox-collection

pfefferle committed:
move map functions to base transformer

pfefferle committed:
use item instead of wp_object

pfefferle committed:
protect functions

pfefferle committed:
fix tests and remove legacy tests

pfefferle committed:
fix phpcs

pfefferle committed:
add unittests

pfefferle committed:
fix attachment issues and add tests

pfefferle committed:
add tests for JSON transformer and fix detected problems

pfefferle committed:
fix phpcs issues

pfefferle committed:
add activity-object transformer tests

pfefferle committed:
simplify code

pfefferle committed:
define as global functions

pfefferle committed:
re-added setter

pfefferle committed:
Merge branch 'trunk' into add/outbox-collection

pfefferle committed:
add content visibility

pfefferle committed:
fix PHPCS

obenland committed:
Outbox: Use post meta instead of taxonomies (#1173)

* Outbox: Use post meta instead of taxonomies

* Add tests for meta values when adding Outbox items

* fix indents

---------

Co-authored-by: Matthias Pfefferle <pfefferle@users.noreply.github.com>

pfefferle committed:
Update includes/scheduler/class-actor.php

Co-authored-by: Matt Wiebe <wiebe@automattic.com>

pfefferle committed:
register content-visibility meta

pfefferle committed:
rename dispatcher

pfefferle committed:
rename dispatcher

pfefferle committed:
Simple dispatcher based on the Outbox-Collection

This is a simple rewrite of the current dispatcher system, to use the Outbox instead of the Scheduler.

This is a first draft and will be improved over time, to better handle:

* Re-tries
* Errors
* Logging
* Batch processing

pfefferle committed:
Simple dispatcher based on the Outbox-Collection

This is a simple rewrite of the current dispatcher system, to use the Outbox instead of the Scheduler.

This is a first draft and will be improved over time, to better handle:

* Re-tries
* Errors
* Logging
* Batch processing

pfefferle committed:
update changelog

pfefferle committed:
update changelog

pfefferle committed:
Merge branch 'trunk' into add/outbox-collection

obenland committed:
Outbox: Update API Endpoint (#1170)

* Update file name

* Extend WP_REST_Controller

* First pass at updated outbox endpoint

* Remove expectation of having cc in the response

* Update object type to be a bit more descriptive

* First pass at create endpoint

* Fix tests

* Get Activity Type from meta after #1173

* Fix rest_url_path

Props @pfefferle

* Return accurate number of outbox items

Props @pfefferle

* Add more tests and remove unnecessary out-of-bounds error

* Remove POST endpoint for now

We currently don't have a use-case for it.

* Limit Outbox by activity type and visibility

* Account for posts without visibility meta

Props @pfefferle

* phpcs

* Allow logged in users to see all activity types

* Add request context to remaining hooks

* Default query to limit activities

Adds an exception for requests with privileges.

* Slightly improve query

pfefferle committed:
fix unittests

pfefferle committed:
Merge branch 'add/outbox-collection' of https://github.com/Automattic/wordpress-activitypub into add/outbox-collection

pfefferle committed:
Merge branch 'add/outbox-collection' into update/dispatcher

pfefferle committed:
pending seems to be the better status

pfefferle committed:
mark post as `publish` after federation id done

pfefferle committed:
mark post as `publish` after federation id done

pfefferle committed:
show only published activities

pfefferle committed:
show only published activities

pfefferle committed:
fix missing rename

pfefferle committed:
fix missing rename

pfefferle committed:
Revert "pending seems to be the better status"

This reverts commit 8a7b5ccbf3f55bc5ce85b04e7366e214a09d32e0.

pfefferle committed:
use pending instead of draft

pfefferle committed:
use pending instead of draft

mattwiebe committed:
Merge branch 'add/outbox-collection' into update/dispatcher

mattwiebe committed:
Merge branch 'trunk' into add/outbox-collection

mattwiebe committed:
Merge branch 'add/outbox-collection' into update/dispatcher

pfefferle committed:
do not check for post_status

pfefferle committed:
do not check for post_status

pfefferle committed:
fix tests

props @obenland

pfefferle committed:
fix tests

props @obenland

mattwiebe committed:
Send `Update`s to Blog Actor in dual mode

mattwiebe committed:
Send `Update`s to Blog Actor in dual mode

mattwiebe committed:
Add logging

pfefferle committed:
Update includes/class-dispatcher.php

Co-authored-by: Konstantin Obenland <obenland@gmx.de>

pfefferle committed:
Update includes/class-dispatcher.php

Co-authored-by: Konstantin Obenland <obenland@gmx.de>

pfefferle committed:
Update includes/rest/class-outbox-controller.php

Co-authored-by: Konstantin Obenland <obenland@gmx.de>

pfefferle committed:
Check if Activity should be sent to followers

pfefferle committed:
the unique check will be done `send_activity_to_followers`

pfefferle committed:
fix tests

pfefferle committed:
fix PHPCS

pfefferle committed:
move scheduler behind action

pfefferle committed:
Add `private` visibility

pfefferle committed:
Add Announce activity

pfefferle committed:
Announce the full object!

pfefferle committed:
fix indent

pfefferle committed:
Update includes/transformer/class-base.php

Co-authored-by: Konstantin Obenland <obenland@gmx.de>

pfefferle committed:
add doc-block

obenland committed:
Outbox: Use callbacks that can be unhooked. (#1188)

* Outbox: Use callbacks that can be unhooked.

* Add unit tests

* Use data provider in no-activity tests

* Make sure user has activitypub cap

* Use should_be_federated() instead

See #1186.

* Update with learnings from #1196

obenland committed:
Outbox: Make sure Last page is linked correctly (#1195)

Props @pfefferle

pfefferle committed:
only boost content not profile updates

pfefferle committed:
Update includes/collection/class-outbox.php

Co-authored-by: André Menrath <99024746+Menrath@users.noreply.github.com>

pfefferle committed:
Update includes/collection/class-outbox.php

Co-authored-by: André Menrath <99024746+Menrath@users.noreply.github.com>

pfefferle committed:
Update includes/collection/class-outbox.php

Co-authored-by: André Menrath <99024746+Menrath@users.noreply.github.com>

pfefferle committed:
fix type

props @menrath

mattwiebe committed:
Also handle `Delete` when bundling Blog Actor inboxes

obenland commented:
@mattwiebe @pfefferle What do we want to do with the data we log? Is it meant for site owners or us?

Maybe we could start with `error_log()`?
I wonder if it would make sense to integrate with Debug Bar. We could also expand on `Stream_Connector`.

obenland committed:
Update docs

obenland committed:
Avoid activitypub_actor_mode bleeding into other tests

obenland committed:
Fix comments tests

obenland committed:
Account for inheritance in Activity objects

obenland committed:
Move hook to the right place

mattwiebe commented:
> What do we want to do with the data we log? Is it meant for site owners or us?

A few use-cases for site owners:

1. Being able to verify that updates are being POSTed to followers. Most of the work our plugin does is invisible to the user, this can make it visible. That's why I'm storing successful logs, not just errors.
2. Ability to manually retry failed POSTs (would need some UI)
3. Monitoring: show a "networking health" dashboard, highlight servers with failed POSTs

But also for us, give the ability to, say, export a debug log of failed requests (sanitized of any PII) when getting support would give us a huge advantage over today.

> I wonder if it would make sense to integrate with Debug Bar.

That seems like a great idea! I was initially thinking of integrating it into an Outbox post_type listing as a column.

mattwiebe committed:
Merge branch 'update/dispatcher' into add/dispatch-logging

pfefferle committed:
Outbox: Fix Query (#1200)

* Outbox: Fix Query

If `author` is null, it will return every activity for the blog and application Actor, but we only want to have the Actors activities.

* Update includes/rest/class-outbox-controller.php

Co-authored-by: Konstantin Obenland <obenland@gmx.de>

* fix user check!

* fix phpcs

* use constants

* Add tests for new functionality

* Update tests/includes/rest/class-test-outbox-controller.php

---------

Co-authored-by: Konstantin Obenland <obenland@gmx.de>

pfefferle committed:
Fix typo

pfefferle commented on includes/class-dispatcher.php:
```diff
@@ -104,8 +104,24 @@ private static function send_activity_to_followers( $activity, $actor_id, $outbo
 
+		// We will store the json as generated by the transformer, even though it 's also in plaintext in $outbox_item->post_content
+		// This will also allow us to keep the logs below leaner.
+		\add_post_meta( $outbox_item->ID, 'activitypub_sent_json', $json );
```
why do we need this information?

mattwiebe replied:
Coming back to it, I was just thinking of all of the places I would want to possibly `error_log` while debugging this code. And this was one. But since we store it in the outbox content anyway, it's definitely superfluous and I'll remove it.

pfefferle commented on includes/class-dispatcher.php:
```diff
@@ -104,8 +104,24 @@ private static function send_activity_to_followers( $activity, $actor_id, $outbo
+		// This will also allow us to keep the logs below leaner.
+		\add_post_meta( $outbox_item->ID, 'activitypub_sent_json', $json );
+		// This will allow error checking later, that all inboxes have been sent.
+		\add_post_meta( $outbox_item->ID, 'activitypub_sent_inboxes', $inboxes, true );
```
shouldn't we simply keep a list of inboxes that bounce? this is a lot of data we store per Outbox Activity!?!

mattwiebe replied:
I wanted a fully debuggable picture of what happens, given that the list of inboxes POSTed to for any given outbox activity will be somewhat dynamically generated. At the same time I was envisioning using it as a basis for rolling batching, which I did not get to. 

But since we're not doing batching yet either, and all of the inboxes are stored in the logs below, we don't need it right now.

obenland committed:
Outbox: Tests and fixes for Post Scheduler (#1196)

* Outbox: Tests and fixes for Post Scheduler

* Move extrafields post to Actor class

* Add test coverage for remaining Actor methods

* Move cap-manipulating tests to the end

* Account for inheritance in Transformer Factory

Reverts changes to `add_to_outbox()`

* Remove outbox items after every test.

* Use fully qualified class name

* Split extra fields tests to not hit time limit

* Adjust bog modes

* debug

* debug

* debug

* debug

* Use Profile_update action.

`wp_update_user` wasn't introduced until WP 6.3.

* revert changes to add_to_outbox

pfefferle committed:
Update Dispatcher to use Outbox (#1186)

* rename dispatcher

* Simple dispatcher based on the Outbox-Collection

This is a simple rewrite of the current dispatcher system, to use the Outbox instead of the Scheduler.

This is a first draft and will be improved over time, to better handle:

* Re-tries
* Errors
* Logging
* Batch processing

* update changelog

* mark post as `publish` after federation id done

* show only published activities

* fix missing rename

* use pending instead of draft

* do not check for post_status

* fix tests

props @obenland

* Send `Update`s to Blog Actor in dual mode

* Update includes/class-dispatcher.php

Co-authored-by: Konstantin Obenland <obenland@gmx.de>

* Update includes/class-dispatcher.php

Co-authored-by: Konstantin Obenland <obenland@gmx.de>

* Update includes/rest/class-outbox-controller.php

Co-authored-by: Konstantin Obenland <obenland@gmx.de>

* Check if Activity should be sent to followers

* the unique check will be done `send_activity_to_followers`

* fix tests

* fix PHPCS

* move scheduler behind action

* Add `private` visibility

* Add Announce activity

* Announce the full object!

* fix indent

* Update includes/transformer/class-base.php

Co-authored-by: Konstantin Obenland <obenland@gmx.de>

* add doc-block

* only boost content not profile updates

* Also handle `Delete` when bundling Blog Actor inboxes

* Update docs

* Avoid activitypub_actor_mode bleeding into other tests

* Fix comments tests

* Account for inheritance in Activity objects

* Move hook to the right place

* fix typo!

* trigger scheduler

* Fix tests

---------

Co-authored-by: Matt Wiebe <wiebe@automattic.com>
Co-authored-by: Konstantin Obenland <obenland@gmx.de>

mattwiebe committed:
Merge branch 'add/outbox-collection' into add/dispatch-logging

mattwiebe committed:
lighter logging

mattwiebe committed:
cleanup merge mess

pfefferle commented:
Closed in favour of the current batch implementation
